test/core/xds/xds_client_test.cc

```diff
@@ -631,11 +656,25 @@ class XdsClientTest : public ::testing::Test {
       MutexLock lock(&mu_);
       return resource_updates_valid_;
     }
+    const ResourceUpdateMap* resource_updates_valid_ptr() const {
```

Comment:
Exposing all of this internal state is ugly and shouldn't be necessary.  Instead, I suggest something like this:

```
// Returns true if the matcher returns true before the timeout.
// Runs the predicate once as soon as it is called and then again
// every time the metrics reporter sees an update.
bool WaitForMetricsReporterData(
    ::testing::Matcher<ResourceUpdateMap> resource_updates_valid_matcher,
    ::testing::Matcher<ResourceUpdateMap> resource_updates_invalid_matcher,
    ::testing::Matcher<ServerFailureMap> server_failures_matcher,
    absl::Duration timeout = absl::Seconds(3),
    SourceLocation location = SourceLocation()) {
  const absl::Time deadline = absl::Now() + (timeout * grpc_test_slowdown_factor());
  MutexLock lock(&mu_);
  while (true) {
    if (::testing::Matches(resource_updates_valid_matcher)(resource_updates_valid_) &&
        ::testing::Matches(resource_updates_invalid_matcher)(resource_updates_invalid_) &&
        ::testing::Matches(server_failures_matcher)(server_failures_)) {
      return true;
    }
    if (cond_.WaitWithDeadline(&mu_, deadline)) break;
  }
  EXPECT_THAT(resource_updates_valid_, resource_updates_valid_matcher)
      << location.file() << ":" << location.line();
  EXPECT_THAT(resource_updates_invalid_, resource_updates_invalid_matcher)
      << location.file() << ":" << location.line();
  EXPECT_THAT(server_failures_, server_failures_matcher)
      << location.file() << ":" << location.line();
  return false;
}
```

Then the tests can invoke it using something like this:

```
EXPECT_TRUE(metrics_reporter_->WaitForMetricsReporterData(
    ::testing::ElementsAre(::testing::Pair(
        ::testing::Pair(kDefaultXdsServerUrl,
                        XdsFooResourceType::Get()->type_url()),
        1)),
    ::testing::_, ::testing::_));
```